File: Strivers A2Z DSA Sheet/34-FindFirstAndLastPositionOfElementInSortedArray/34-FindFirstAndLastPositionOfElementInSortedArray.cpp
// Last updated: 8/7/2026, 7:17:13 PM
class Solution {
public:

    int firstOcc(vector<int>& nums, int target) {
        int low = 0;
        int high = nums.size() - 1;
        int ans = -1;

        while(low <= high){
            int mid = (low + high)/2;

            if(nums[mid] == target){
                ans = mid;
                high = mid - 1;
            }
            else if(nums[mid] < target) low = mid+1;
            else high = mid - 1;
        }

        return ans;
    }


    int lastOcc(vector<int>& nums, int target) {
        int low = 0;
        int high = nums.size() - 1;
        int ans = nums.size();

        while(low <= high){
            int mid = (low + high)/2;

            if(nums[mid] == target){
                ans = mid;
                low = mid + 1;
            }
            else if(nums[mid] < target) low = mid+1;
            else high = mid - 1;
        }

        return ans;
    }
    vector<int> searchRange(vector<int>& nums, int target) {
        int first = firstOcc(nums, target);
        vector<int> ans(2,-1);
        if(first == -1) return ans;

        ans.clear();
        ans.push_back(first); 
        ans.push_back(lastOcc(nums,target));

        return ans; 
    }
};

File: Strivers A2Z DSA Sheet/34-FindFirstAndLastPositionOfElementInSortedArray/34-FindFirstAndLastPositionOfElementInSortedArray.java
// Last updated: 8/7/2026, 7:17:08 PM
class Solution {
    public static int firstOcc(int nums[], int target){
        int low = 0;
        int high = nums.length-1;
        int ans = -1;
        while(low<=high){
            int mid = low + (high-low)/2;

            if(nums[mid] == target){
                ans = mid;
                high = mid-1;
            }
            else if(nums[mid]<target) low = mid+1;
            else high = mid-1;
        }
        return ans;
    }

    public static int lastOcc(int nums[], int target){
        int low = 0;
        int high = nums.length-1;
        int ans = high;
        while(low<=high){
            int mid = low + (high-low)/2;

            if(nums[mid] == target){
                ans = mid;
                low = mid+1;
            }
            else if(nums[mid]>target) high = mid-1;
            else low = mid+1;
        }
        return ans;
    }

    public int[] searchRange(int[] nums, int target) {
        int first = firstOcc(nums, target);
        int[] ans = {-1, -1};
        if(first == -1) return ans;

        ans[0] = first;
        ans[1] = lastOcc(nums, target);
        return ans;
    }
}